### ISSUE TITLE
adding type annotation 

### ISSUE THREAD

**Issue by martinResearch** [uqfoundation/multiprocess#128]
Thank you for the library.
Unfortunately, having this library not typed prevents mypy from propagating types to do efficient static checks.
Is adding type annotation in the roadmap?
Maybe the stubs in here https://github.com/python/typeshed/tree/main/stdlib/multiprocessing could be reused?

**Comment by mmckerns** [comment]
No, adding type annotations is not currently in the roadmap. Thanks for the pointer to `multiprocessing` from `typeshed`.  I'll have to look into whether it's realistic to use that as a base as opposed to the stdlib `multiprocessing`.

**Comment by martinResearch** [comment]
To be sure there is no confusion: It seems to me that we cannot really oppose the code in here https://github.com/python/cpython/blob/main/Lib/multiprocessing with the code in here https://github.com/python/typeshed/tree/main/stdlib/multiprocessing. It seems that the first one contains the library implementation and does not have any type annotation while the second contains only the annotations and no implementation.

**Comment by mmckerns** [comment]
👍 there's no confusion

**Comment by Jylpah** [comment]
Second vote for type hints. Type hints and `mypy` make Python programming for more robust since the linter will notice many (most of) issues that would otherwise be runtime errors. But I can image it would be quite an effort to add.

**Comment by mcrespoae** [comment]
Following up on this issue: Third vote for type hints.

**Comment by chdominguez** [comment]
As a workaround, you can cast `multiprocess` as `multiprocessing` for type annotations during development using the typing module:

```
if typing.TYPE_CHECKING:
    import multiprocessing as mp
else:
    import multiprocess as mp
```

or 

```
import multiprocess as mp
if typing.TYPE_CHECKING:
   import multiprocessing
   mp = typing.Cast(multiprocessing, mp)
```

**Comment by zhou13** [comment]
I created a stubs package so it works for me: https://github.com/zhou13/multiprocess-stubs.

**Comment by codethief** [comment]
@chdominguez I believe in your second workaround there's a typo: `Cast` should be `cast`.

Either way, for some weird reason it doesn't seem to work in Pylance/Pyright, unfortunately: In expressions like `mp.Process()` go-to-definition in VSCode will work (well, it will take you to `multiprocessing.Process` instead of `multiprocess.Process`) but Pylance will still underline `Process` and, likewise, Pyright will complain on the command line:

```
foo.py:34:22 - error: "Process" is not a known attribute of module "multiprocess"
```

Meanwhile, your first workaround works without issues.

**Comment by zhou13** [comment]
@codethief Yes, I simply copied files from https://github.com/python/typeshed/tree/main/stdlib/multiprocessing. Do you still have issues using it?

**Comment by codethief** [comment]
@zhou13 My apologies for the delay, the notification about your message must have gotten burried.

> Do you still have issues using it?

My point was that your stubs would probably be better if they referenced `multiprocess` instead of `multiprocessing`. (You can probably just run a search & replace.) Otherwise, like in @chdominguez's workaround, type checkers will validate usages against `multiprocessing`[0], and go-to-definition (ctrl+click) for e.g. `multiprocess.Process` will take you to the source code of `multiprocessing.Process`.

[0]: Obviously, this is fine as long as the interfaces of `multiprocess` and `multiprocessing` are *exactly* the same.
